core/species.cpp

```diff
@@ -5623,11 +6272,24 @@ void Species::FreeTreeSequence()
 	{
 		// Free any tree-sequence recording stuff that has been allocated; called when Species is getting deallocated,
 		// and also when we're wiping the slate clean with something like readFromPopulationFile().
-		tsk_table_collection_free(&tables_);
-		tables_initialized_ = false;
+		bool first = true;
+		for (TreeSeqInfo &tsinfo : treeseq_)
+		{
+			// the node, individual, and population tables are shared; avoid doing a double free
+			// (I don't think any of the shared tables should be copied at this point anyway,
+			// though; maybe there should be an assert here to that effect?)
```

Comment:
looks good but FYI `tsk_table_collection_free` uses [`tsk_safe_free`](https://github.com/tskit-dev/tskit/blob/main/c/tskit/core.c#L726), which is "safe" because it won't try to free a NULL pointer

Reply:
Ah, true.  Jerome's code in the example had the same "avoid a double free", so I just copied his approach, but perhaps it is technically unnecessary.

Reply:
(and technically free() is safe to call with NULL anyway – https://stackoverflow.com/a/6084233/2752221)

Reply:
Ah, I see why Jermome did this, and it is needed.  Each table collection its own copy of the pointers for columns.  So `tsk_safe_free()` doesn't protect this case.  You'd free the first table collection, would would zero out its column pointers, so if you tried to free _that_ table collection again, it would be OK – but the _other_ table collections still have non-zeroed pointers to the same memory that just got freed, and calling `tsk_safe_free()` on those stale pointers would still go boom.  Resolving this thread now (hopefully you're seeing these comments, even though I'm resolving!)